### PR TITLE
chore: enhance useTrackedEffect type

### LINKED ISSUE #2195: The argments's type of useTrackedEffect's effect function can not be accurately deduced.
The argments's type of useTrackedEffect's effect function can not be accurately deduced.

Example:

`
useTrackedEffect((changes, preDeps, deps) => {}, ['abc', 123]);
`

Expect `preDeps`&`deps` type will be `[string, number]`, but get `any[]`.

### PR DESCRIPTION
**Pull request by xvxlb**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
fix: #2195 

### 💡 Background and solution
#2195 

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | chore: enhance useTrackedEffect type |
| 🇨🇳 Chinese | chore: useTrackedEffect 类型修改 |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: packages/hooks/src/useTrackedEffect/index.ts [hunk 1]
```diff
@@ -20,8 +20,8 @@ const diffTwoDeps = (deps1?: DependencyList, deps2?: DependencyList) => {
     : [];
 };
 
-const useTrackedEffect = (effect: Effect, deps?: DependencyList) => {
-  const previousDepsRef = useRef<DependencyList>();
+const useTrackedEffect = <T extends any[]>(effect: Effect<T>, deps?: [...T]) => {
+  const previousDepsRef = useRef<T>();
```
Comment (liuyib) [inline packages/hooks/src/useTrackedEffect/index.ts:24]: how about replacing `T extends any[]` with `T extends unknown[]`?

Comment (xvxlb) [reply packages/hooks/src/useTrackedEffect/index.ts:24]: `unknown` is better and code is updated.

**Comment by liuyib** [comment]
I find a better way to do this:

```diff
import type { DependencyList } from 'react';
import { useEffect, useRef } from 'react';

- type Effect<T extends unknown[]> = (
+ type Effect<T extends DependencyList> = (
   changes?: number[],
-  previousDeps?: [...T],
+  previousDeps?: T,
-  currentDeps?: [...T],
+  currentDeps?: T,
) => void | (() => void);

const diffTwoDeps = (deps1?: DependencyList, deps2?: DependencyList) => {
  //Let's do a reference equality check on 2 dependency list.
  //If deps1 is defined, we iterate over deps1 and do comparison on each element with equivalent element from deps2
  //As this func is used only in this hook, we assume 2 deps always have same length.
  return deps1
    ? deps1
        .map((_ele, idx) => (!Object.is(deps1[idx], deps2?.[idx]) ? idx : -1))
        .filter((ele) => ele >= 0)
    : deps2
    ? deps2.map((_ele, idx) => idx)
    : [];
};

- const useTrackedEffect = <T extends unknown[]>(effect: Effect<T>, deps?: [...T]) => {
+ const useTrackedEffect = <T extends DependencyList>(effect: Effect<T>, deps?: [...T]) => {
  const previousDepsRef = useRef<T>();

  useEffect(() => {
    const changes = diffTwoDeps(previousDepsRef.current, deps);
    const previousDeps = previousDepsRef.current;
    previousDepsRef.current = deps;
    return effect(changes, previousDeps, deps);
  }, deps);
};

export default useTrackedEffect;
```

I think the built-in type `DependencyList` is better than `unknown[]`, what do you think? @xvxlb

**Comment by xvxlb** [comment]
> 

@liuyib This seems to be more accurate 👍

**Review by liuyib — APPROVED** [review]
Nice work, thx

**Comment by xvxlb** [comment]
@liuyib need merge.

**Comment by liuyib** [comment]
> @liuyib need merge.

I don't have permission to merge :)

@crazylxr review, pls

**Review by crazylxr — APPROVED** [review]

**Pull request merged**